bundles/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/internal/GardenaSmartWebSocket.java

```diff
@@ -97,34 +90,38 @@ public synchronized void stop() {
         if (connectionTracker != null) {
             connectionTracker.cancel(true);
         }
-        if (isRunning()) {
-            logger.debug("Closing Gardena Webservice client ({})", socketId);
-            try {
-                session.close();
-            } catch (Exception ex) {
-                // ignore
-            } finally {
-                try {
-                    webSocketClient.stop();
-                } catch (Exception e) {
-                    // ignore
-                }
-            }
+
+        logger.debug("Closing Gardena Webservice ({})", socketId);
+        try {
+            session.close();
+        } catch (Exception ex) {
+            // ignore
         }
     }
 
-    /**
-     * Returns true, if the websocket is running.
-     */
-    public synchronized boolean isRunning() {
-        return session.isOpen();
+    public boolean isClosing() {
+        return this.closing;
+    }
+
+    public String getSocketID() {
+        return this.socketId;
+    }
+
+    public String getLocationID() {
+        return this.locationID;
+    }
+
+    public void restart(String newUrl) throws Exception {
+        logger.debug("Reconnecting to Gardena Webservice ({})", socketId);
+        session = (WebSocketSession) webSocketClient.connect(this, new URI(newUrl)).get();
     }
 
     @OnWebSocketConnect
     public void onConnect(Session session) {
         closing = false;
         logger.debug("Connected to Gardena Webservice ({})", socketId);
 
+        // start sending PING every two minutes
         connectionTracker = scheduler.scheduleWithFixedDelay(this::sendKeepAlivePing, 2, 2, TimeUnit.MINUTES);
```

Comment:
I have to disagree, that this is a "major" memory leak. Currently it is guaranteed that the `stop()` function is called before every `restart()` in case there is some kind of error. See `GardenaSmartImpl.restartWebsocket(@Nullable GardenaSmartWebSocket socket)`.

But I can still add your suggestion, so that we are safe if the code gets changed at some point in time.

Reply:
This line must become `this.connectionTracker = scheduler...` though. At the moment you're assigning to the local variable `connectionTracker`, which leads to `this.connectionTracker` never becoming non-null.

Reply:
Sorry about that. It was definitely too late for me yesterday...